<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Collection of functions created and/or curated to aid in the visualization and analysis of single-cell data using R. • scCustomize</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Collection of functions created and/or curated to aid in the visualization and analysis of single-cell data using R.">
<meta property="og:description" content="Custom functions to modify single cell analyses in publically available packages.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">scCustomize</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.0.9998</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">Getting Started</a>
</li>
<li>
  <a href="articles/13_Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Plotting Vignettes</li>
    <li>
      <a href="articles/01_Gene_plotting.html">Plotting #1: Analysis Plots</a>
    </li>
    <li>
      <a href="articles/02_QC_Plots.html">Plotting #2: QC Plots</a>
    </li>
    <li>
      <a href="articles/03_Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    </li>
    <li>
      <a href="articles/04_Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    </li>
    <li>
      <a href="articles/05_Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Helpers &amp; Utilities</li>
    <li>
      <a href="articles/06_Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    </li>
    <li>
      <a href="articles/07_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    </li>
    <li>
      <a href="articles/08_Statistics.html">Statistics Functions</a>
    </li>
    <li>
      <a href="articles/09_Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    </li>
    <li>
      <a href="articles/12_Misc_Functions.html">Misc Functions</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other Package Interactivity/Functions</li>
    <li>
      <a href="articles/10_LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    </li>
    <li>
      <a href="articles/11_Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">News/Changelog</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/samuel-marsh/scCustomize/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="getting-stated-with-sccustomize-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#getting-stated-with-sccustomize-" class="anchor"></a>Getting Stated with scCustomize <img src="data/scCustomize_Logo.svg" align="right" width="150">
</h1></div>
<p><img src="https://img.shields.io/github/v/release/samuel-marsh/scCustomize?style=flat-square" alt="version"> [<img src="https://img.shields.io/github/license/samuel-marsh/scCustomize?style=flat-square" alt="license"><a href="https://github.com/samuel-marsh/scCustomize/issues"><img src="https://img.shields.io/github/issues/samuel-marsh/scCustomize?style=flat-square" alt="issues"></a> <img src="https://img.shields.io/github/downloads/samuel-marsh/scCustomize/total?style=flat-square" alt="GitHub all releases"></p>
<p>scCustomize is a collection of functions created and/or curated to aid in the visualization and analysis of single-cell data using R.</p>
<div id="installing-sccustomize" class="section level2">
<h2 class="hasAnchor">
<a href="#installing-sccustomize" class="anchor"></a>Installing scCustomize</h2>
<p>Please see <a href="LINK">Installation</a> page for full installation instructions and GitHub branch information.</p>
</div>
<div id="vignettestutorials" class="section level2">
<h2 class="hasAnchor">
<a href="#vignettestutorials" class="anchor"></a>Vignettes/Tutorials</h2>
<p>See Vignettes for detailed tutorials of all aspects of scCustomize functionality.</p>
</div>
<div id="goalsabout-sccustomize" class="section level2">
<h2 class="hasAnchor">
<a href="#goalsabout-sccustomize" class="anchor"></a>Goals/About scCustomize</h2>
<p><strong><em>The goals of scCustomize are to:</em></strong></p>
<p><strong><em>1. Customize visualizations for both ease of use and more aesthetic visuals.</em></strong><br><strong><em>2. Improve speed/reproducibility of common tasks/pieces of code in scRNA-seq analysis with a single or group of functions.</em></strong></p>
<p>scCustomize aims to acheive these goals through:</p>
<ul>
<li>
<strong>Customized versions of many commonly used plotting functions (and some custom ones).</strong><br>
To create greater flexibility in visualization and more aesthetic visuals by:
<ul>
<li>Altering default parameters for more intuitive plots (or at least I believe plots). For instance: <code>FeaturePlot(..., order = TRUE)</code>.<br>
</li>
<li>Wrapping commonly used ggplot2 post-plot themeing into function call. No more copy/paste of the same theme elements for every plot over and over (e.g., <code>plot + scale_color_continuous(...) + ggtitle(...) + theme(plot.title = element_text(...), legend.position = ...) + guides(...)</code><br>
</li>
<li>Creating new plotting functions either: 1. as wrapper around Seurat function with parameters already specified (e.g., <code>QC_Plot_Genes</code>) or 2. create new plots (e.g., <code>Seq_QC_Plot_Reads_per_Cell</code> or <code>Plot_Median_Genes</code>) or 3. both (e.g., <code>QC_Plot_UMIvsGene(..., combination = TRUE)</code>).<br>
</li>
<li>Adding additional parameters to existing plots inside new function (e.g., high and low cutoff parameters in <code>QC_Plot_UMIvsGene</code>)</li>
</ul>
</li>
<li>
<strong>Easy iterative plotting functionality.</strong><br>
Many plotting functions can be easily automated with loops, apply, purrr etc. However, these can be intimidating to novice user and often can be made easier through wrapping into a function.
<ul>
<li>scCustomize contains a number of iterative functions contain extra parameters to specify file type, path, name and then render progress bar in console to track progress.</li>
<li>Returns either single PDF document or multiple plots of any valid file type (e.g., png, tiff, jpeg, pdf, etc).</li>
</ul>
</li>
<li>
<strong>Helper functions easily import multiple raw data types</strong><br>
Functions are aimed at streamlining importing multiple files/samples with single function and/or importing files with “non-standard” file names.
<ul>
<li>Iterate import functions to simplify the import process across groups of files/samples.</li>
<li>Parallelize functions where possible to allow for dramatic speed improvements when import large number of samples simultaenously.</li>
</ul>
</li>
<li>
<strong>Helper functions to simplify analysis with addition of new default parameters or wrapping multiple lines of code into single function.</strong><br>
Goal is to both speed up and simplify coding and reduce the use of copy/paste of the same lines of code which is more likely to lead to errors in code reproducibility.
<ul>
<li>
<p><em>Example of adding new parameters:</em> Adding the percentage of counts aligning to mitocondrial (and/or ribosomal) genes is common early step in analysis. scCustomize provides <code>Add_Mito_Ribo_Seurat</code> (and LIGER version) to simplify this. Basic use requires only one line of code and two parameters.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Add_Mito_Ribo_Seurat</span>(<span class="at">seurat_object =</span> obj_name, <span class="at">species =</span> <span class="st">"Human"</span>) </span></code></pre></div>
<ul>
<li>Function already knows the defaults for Human, Mouse, and Marmoset (submit a PR if you would like more species added!).</li>
</ul>
</li>
<li>
<p><em>Example of wrapping many lines to one:</em> Extracting the top 10 (or 15, 20, 25, etc) genes per identity after running <code>FindAllMarkers</code> is very common and scCustomize provides <code>Extract_Top_Markers</code> function to simplify process.<br><em>Using scCustomize function:</em></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>markers_df <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(<span class="at">object =</span> obj_name)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Get vector/string</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>top10_list <span class="ot">&lt;-</span> <span class="fu">Extract_Top_Markers</span>(<span class="at">marker_dataframe =</span> markers_df)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># or for data.frame</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>top10_df <span class="ot">&lt;-</span> <span class="fu">Extract_Top_Markers</span>(<span class="at">marker_dataframe =</span> markers_df, <span class="at">dataframe =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><em>Instead of tidyverse/base R:</em></p>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>markers_df <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(<span class="at">object =</span> obj_name)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Get vector/string</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>top10_list <span class="ot">&lt;-</span> markers_df <span class="sc">%&gt;%</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">"rownames"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(<span class="at">n =</span> <span class="dv">10</span>, <span class="at">order_by =</span> avg_log2FC) <span class="sc">%&gt;%</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(<span class="st">"gene"</span>)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># or for data.frame</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>top10_df <span class="ot">&lt;-</span> markers_df <span class="sc">%&gt;%</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">"rownames"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(<span class="at">n =</span> <span class="dv">10</span>, <span class="at">order_by =</span> avg_log2FC) <span class="sc">%&gt;%</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"rownames"</span>)</span></code></pre></div>
</li>
</ul>
</li>
<li>
<strong>Provide more informative error messages for many common issues</strong><br>
Base R error messages resutling from error deep inside Seurat (or other package) function can sometimes be difficult to interpret, especially for userse new to R.
<ul>
<li>scCustomize provides checks/warnings wrapped inside its functions to help and provide more informative error/warning messages. Two examples include:<br>
</li>
<li>
<code>Add_Mito_Ribo_Seurat</code> will warn you if no mitochondrial or ribosomal features are found and won’t create new metadata column.<br>
</li>
<li>
<code>Rename_Clusters</code> will check and make sure the right number of unique new names are provided and provide one of two error messages if not before attempting to rename the object idents.</li>
</ul>
</li>
</ul>
</div>
<div id="support-for-other-scrna-seq-object-formats-liger-sce-etc" class="section level2">
<h2 class="hasAnchor">
<a href="#support-for-other-scrna-seq-object-formats-liger-sce-etc" class="anchor"></a>Support for Other scRNA-seq Object Formats (LIGER, SCE, etc)</h2>
<p>Currently the package is primarily centered around interactivity with Seurat Objects with some functionality with LIGER objects.<br>
If users are interested in adapting functions (or creating separate functions) to provide comparable functionality with SCE or other object formats I would be happy to add them. See below for more info on PRs.</p>
</div>
<div id="bug-reportsnew-features" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-reportsnew-features" class="anchor"></a>Bug Reports/New Features</h2>
<div id="if-you-run-into-any-issues-or-bugs-please-submit-a-github-issue-with-details-of-the-issue" class="section level4">
<h4 class="hasAnchor">
<a href="#if-you-run-into-any-issues-or-bugs-please-submit-a-github-issue-with-details-of-the-issue" class="anchor"></a>If you run into any issues or bugs please submit a <a href="https://github.com/samuel-marsh/scCustomize/issues">GitHub issue</a> with details of the issue.</h4>
<ul>
<li>If possible please include a reproducible example (suggest using <a href="https://github.com/satijalab/seurat-data">SeuratData package</a> pbmc dataset for lightweight examples.)</li>
</ul>
</div>
<div id="any-requests-for-new-features-or-enhancements-can-also-be-submitted-as-github-issues" class="section level4">
<h4 class="hasAnchor">
<a href="#any-requests-for-new-features-or-enhancements-can-also-be-submitted-as-github-issues" class="anchor"></a>Any requests for new features or enhancements can also be submitted as <a href="https://github.com/samuel-marsh/scCustomize/issues">GitHub issues</a>.</h4>
<ul>
<li>Even if you don’t know how to implement/incorporate with current package go ahead a submit!</li>
</ul>
</div>
<div id="pull-requests-are-welcome-for-bug-fixes-new-features-or-enhancements" class="section level4">
<h4 class="hasAnchor">
<a href="#pull-requests-are-welcome-for-bug-fixes-new-features-or-enhancements" class="anchor"></a><a href="https://github.com/samuel-marsh/scCustomize/pulls">Pull Requests</a> are welcome for bug fixes, new features, or enhancements.</h4>
<ul>
<li>Please set PR to merge with “develop” branch and provide description of what the PR contains (referencing existing issue(s) if appropriate).</li>
</ul>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/samuel-marsh/scCustomize/">https://​github.com/​samuel-marsh/​scCustomize/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/samuel-marsh/scCustomize/issues">https://​github.com/​samuel-marsh/​scCustomize/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3) | file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Samuel Marsh <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-3012-6945" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
