<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Cell Bender Functionality &amp; Plotting • scCustomize</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Cell Bender Functionality &amp; Plotting">
<meta property="og:description" content="scCustomize">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scCustomize</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Plotting Vignettes</li>
    <li>
      <a href="../articles/Gene_Expression_Plotting.html">Plotting #1: Analysis Plots</a>
    </li>
    <li>
      <a href="../articles/QC_Plots.html">Plotting #2: QC Plots</a>
    </li>
    <li>
      <a href="../articles/Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    </li>
    <li>
      <a href="../articles/Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Helpers &amp; Utilities</li>
    <li>
      <a href="../articles/Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    </li>
    <li>
      <a href="../articles/Markers_and_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    </li>
    <li>
      <a href="../articles/Statistics.html">Statistics Functions</a>
    </li>
    <li>
      <a href="../articles/Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    </li>
    <li>
      <a href="../articles/Misc_Functions.html">Misc Functions</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Package Interactivity/Functions</li>
    <li>
      <a href="../articles/LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    </li>
    <li>
      <a href="../articles/Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/samuel-marsh/scCustomize/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Cell_Bender_Functions_files/kePrint-0.0.1/kePrint.js"></script><link href="Cell_Bender_Functions_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Cell Bender Functionality &amp; Plotting</h1>
            
            <h4 data-toc-skip class="date">Compiled: December 16, 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/samuel-marsh/scCustomize/blob/HEAD/vignettes/Cell_Bender_Functions.Rmd" class="external-link"><code>vignettes/Cell_Bender_Functions.Rmd</code></a></small>
      <div class="hidden name"><code>Cell_Bender_Functions.Rmd</code></div>

    </div>

    
    
<hr>
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level1">
<h1 id="cell-bender-functionality">Cell Bender Functionality<a class="anchor" aria-label="anchor" href="#cell-bender-functionality"></a>
</h1>
<p>Cell Bender is software for elimination of technical artifacts in scRNA-seq/snRNA-seq that uses deep generative model for unsupervised removal of ambient RNA and chimeric PCR artifacts. You can find out more info about Cell Bender from the <a href="https://www.biorxiv.org/content/10.1101/791699v1" class="external-link">bioRxiv Preprint</a>, <a href="https://github.com/broadinstitute/CellBender" class="external-link">GitHub Repo</a>, and <a href="https://cellbender.readthedocs.io/en/latest/" class="external-link">Documentation</a>.</p>
<p>Following completion of Cell Bender scCustomize contains a couple of functions that may be helpful when creating and visualizing data in Seurat.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/samuel-marsh/scCustomize" class="external-link">scCustomize</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/traversc/qs" class="external-link">qs</a></span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="importing-cell-bender-h5-outputs">Importing Cell Bender H5 Outputs<a class="anchor" aria-label="anchor" href="#importing-cell-bender-h5-outputs"></a>
</h2>
<p>The output from Cell Bender is an H5 file that is styled and can be read like 10X Genomics H5 file. However, <code><a href="https://satijalab.org/seurat/reference/Read10X_h5.html" class="external-link">Seurat::Read10X_h5()</a></code> assumes that the file contains no name prefix.</p>
<p>However,<code><a href="../reference/Read10X_h5_Multi_Directory.html">Read10X_h5_Multi_Directory()</a></code> can be used when reading in Cell Bender files. and contains additional parameters related to Cell Bender imports.</p>
<div class="section level3">
<h3 id="cell-bender-output-structure">Cell Bender Output Structure<a class="anchor" aria-label="anchor" href="#cell-bender-output-structure"></a>
</h3>
<p>The following is typical output directory structure for Cell Bender files with sub-directory labeled with sample name and each file also prefixed with the sample name.</p>
<pre><code>Parent_Directory
├── sample_01
│   └── sample_01_out_cell_barcodes.csv
│   └── sample_01_out_filtered.h5
│   └── sample_01_out.h5
│   └── sample_01_out.log
│   └── sample_01_out.pdf
└── sample_02
│   └── sample_02_out_cell_barcodes.csv
│   └── sample_02_out_filtered.h5
│   └── sample_02_out.h5
│   └── sample_02_out.log
│   └── sample_02_out.pdf</code></pre>
</div>
<div class="section level3">
<h3 id="read-in-h5-outputs">Read in H5 outputs<a class="anchor" aria-label="anchor" href="#read-in-h5-outputs"></a>
</h3>
<p>All we have to do is adjust the parameters to account for cell bender file names and directory structure.</p>
<ul>
<li>
<code>secondary_path = ""</code> as the files are directly within the immediate sub-directory</li>
<li>
<code>default_10X_path = FALSE</code> because these are not 10X outputs.</li>
<li>
<code>h5_filename = "_out.h5"</code> specifies the shared aspect of file name that is not part of sample name. (Can also specify “_out_filtered.h5” depending on which file is desired).</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cell_bender_merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Read10X_h5_Multi_Directory.html">Read10X_h5_Multi_Directory</a></span><span class="op">(</span>base_path <span class="op">=</span> <span class="st">"assets/cell_bender/"</span>, secondary_path <span class="op">=</span> <span class="st">""</span>,
    default_10X_path <span class="op">=</span> <span class="cn">FALSE</span>, h5_filename <span class="op">=</span> <span class="st">"_out.h5"</span>, cell_bender <span class="op">=</span> <span class="cn">TRUE</span>, merge <span class="op">=</span> <span class="cn">TRUE</span>, sample_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"WT1"</span>,
        <span class="st">"WT2"</span><span class="op">)</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span>, num_cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="these-matrices-can-then-be-used-to-create-seurat-objects-as-desired-can-also-be-used-to-create-other-analysis-objects">These matrices can then be used to create Seurat objects as desired (can also be used to create other analysis objects)<a class="anchor" aria-label="anchor" href="#these-matrices-can-then-be-used-to-create-seurat-objects-as-desired-can-also-be-used-to-create-other-analysis-objects"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cell_bender_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">cell_bender_merged</span>, names.field <span class="op">=</span> <span class="fl">1</span>, names.delim <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="creating-dual-assay-objects">Creating Dual Assay Objects<a class="anchor" aria-label="anchor" href="#creating-dual-assay-objects"></a>
</h2>
<p>Sometimes it can be helpful to create object that contains both the cell ranger values and cell bender values (we’ll come to why below). scCustomize contains a helper function <code><a href="../reference/Create_CellBender_Merged_Seurat.html">Create_CellBender_Merged_Seurat()</a></code> to handle object creation in one quick step.</p>
<p>For this function we assume that we will use the cell calling algorithm of Cell Ranger with the modified counts for Cell Bender.</p>
<div class="section level3">
<h3 id="read-in-both-sets-of-data">Read in both sets of data<a class="anchor" aria-label="anchor" href="#read-in-both-sets-of-data"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cell_bender_merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Read10X_h5_Multi_Directory.html">Read10X_h5_Multi_Directory</a></span><span class="op">(</span>base_path <span class="op">=</span> <span class="st">"assets/cell_bender/"</span>, secondary_path <span class="op">=</span> <span class="st">""</span>,
    default_10X_path <span class="op">=</span> <span class="cn">FALSE</span>, h5_filename <span class="op">=</span> <span class="st">"_out.h5"</span>, cell_bender <span class="op">=</span> <span class="cn">TRUE</span>, merge <span class="op">=</span> <span class="cn">TRUE</span>, sample_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"WT1"</span>,
        <span class="st">"WT2"</span><span class="op">)</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span>, num_cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>

<span class="va">cell_ranger_merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Read10X_h5_Multi_Directory.html">Read10X_h5_Multi_Directory</a></span><span class="op">(</span>base_path <span class="op">=</span> <span class="st">"assets/cell_ranger/"</span>, default_10X_path <span class="op">=</span> <span class="cn">TRUE</span>,
    h5_filename <span class="op">=</span> <span class="st">"filtered_feature_bc_matrix.h5"</span>, cell_bender <span class="op">=</span> <span class="cn">F</span>, merge <span class="op">=</span> <span class="cn">TRUE</span>, sample_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"WT1"</span>,
        <span class="st">"WT2"</span><span class="op">)</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span>, num_cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-dual-assay-seurat-object">Create Dual Assay Seurat Object<a class="anchor" aria-label="anchor" href="#create-dual-assay-seurat-object"></a>
</h3>
<p>To run the function the user simply needs to provide the names of the two matrices and a name for assay containing the Cell Ranger counts (by default this is named “RAW”).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dual_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Create_CellBender_Merged_Seurat.html">Create_CellBender_Merged_Seurat</a></span><span class="op">(</span>raw_cell_bender_matrix <span class="op">=</span> <span class="va">cell_bender_merged</span>, raw_counts_matrix <span class="op">=</span> <span class="va">cell_ranger_merged</span>,
    raw_assay_name <span class="op">=</span> <span class="st">"RAW"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="optional-parameters">Optional Parameters<a class="anchor" aria-label="anchor" href="#optional-parameters"></a>
</h3>
<p>Users can specify any additional parameters normally passed to <code><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">Seurat::CreateSeuratObject()</a></code> when using this function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dual_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Create_CellBender_Merged_Seurat.html">Create_CellBender_Merged_Seurat</a></span><span class="op">(</span>raw_cell_bender_matrix <span class="op">=</span> <span class="va">cell_bender_merged</span>, raw_counts_matrix <span class="op">=</span> <span class="va">cell_ranger_merged</span>,
    raw_assay_name <span class="op">=</span> <span class="st">"RAW"</span>, min_cells <span class="op">=</span> <span class="fl">5</span>, min_features <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="prepost-cell-bender-analysis">Pre/Post Cell Bender Analysis<a class="anchor" aria-label="anchor" href="#prepost-cell-bender-analysis"></a>
</h2>
<p>It can be very important with tools like Cell Bender to analyze how much the process has effected data on a per cell basis.</p>
<div class="section level3">
<h3 id="add-prepost-to-meta-data">Add Pre/Post to Meta Data<a class="anchor" aria-label="anchor" href="#add-prepost-to-meta-data"></a>
</h3>
<p>scCustomize includes function <code><a href="../reference/Add_Cell_Bender_Diff.html">Add_Cell_Bender_Diff()</a></code> to help with this process. This function will take the nCount and nFeature statistics from both assays in the object and calculate the difference and return 2 new columns (“nCount_Diff” and “nFeature_Diff”) to the object meta.data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">astrocytes_cortex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Cell_Bender_Diff.html">Add_Cell_Bender_Diff</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">astrocytes_cortex</span>, raw_assay_name <span class="op">=</span> <span class="st">"RAW"</span>,
    cell_bender_assay_name <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">astrocytes_cortex</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">5</span><span class="op">)</span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
orig.ident
</th>
<th style="text-align:right;">
nCount_RNA
</th>
<th style="text-align:right;">
nFeature_RNA
</th>
<th style="text-align:right;">
nCount_RAW
</th>
<th style="text-align:right;">
nFeature_RAW
</th>
<th style="text-align:right;">
percent_mito
</th>
<th style="text-align:right;">
percent_ribo
</th>
<th style="text-align:right;">
percent_mito_ribo
</th>
<th style="text-align:right;">
nCount_SCT
</th>
<th style="text-align:right;">
nFeature_SCT
</th>
<th style="text-align:left;">
SCT_snn_res.0.6
</th>
<th style="text-align:left;">
seurat_clusters
</th>
<th style="text-align:right;">
nFeature_Diff
</th>
<th style="text-align:right;">
nCount_Diff
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
WT-2a_AAACCCACACACCAGC-1
</td>
<td style="text-align:left;">
WT-2a
</td>
<td style="text-align:right;">
3207
</td>
<td style="text-align:right;">
1699
</td>
<td style="text-align:right;">
3472
</td>
<td style="text-align:right;">
1852
</td>
<td style="text-align:right;">
0.0311818
</td>
<td style="text-align:right;">
0.0311818
</td>
<td style="text-align:right;">
0.0623636
</td>
<td style="text-align:right;">
3103
</td>
<td style="text-align:right;">
1695
</td>
<td style="text-align:left;">
2
</td>
<td style="text-align:left;">
2
</td>
<td style="text-align:right;">
153
</td>
<td style="text-align:right;">
265
</td>
</tr>
<tr>
<td style="text-align:left;">
WT-2a_AAACGAAGTATCGCTA-1
</td>
<td style="text-align:left;">
WT-2a
</td>
<td style="text-align:right;">
2089
</td>
<td style="text-align:right;">
1217
</td>
<td style="text-align:right;">
2326
</td>
<td style="text-align:right;">
1354
</td>
<td style="text-align:right;">
0.0000000
</td>
<td style="text-align:right;">
0.0478698
</td>
<td style="text-align:right;">
0.0478698
</td>
<td style="text-align:right;">
2468
</td>
<td style="text-align:right;">
1216
</td>
<td style="text-align:left;">
2
</td>
<td style="text-align:left;">
2
</td>
<td style="text-align:right;">
137
</td>
<td style="text-align:right;">
237
</td>
</tr>
<tr>
<td style="text-align:left;">
WT-2a_AAACGAATCGACCATA-1
</td>
<td style="text-align:left;">
WT-2a
</td>
<td style="text-align:right;">
3327
</td>
<td style="text-align:right;">
1677
</td>
<td style="text-align:right;">
3687
</td>
<td style="text-align:right;">
1882
</td>
<td style="text-align:right;">
0.2103998
</td>
<td style="text-align:right;">
0.4207995
</td>
<td style="text-align:right;">
0.6311993
</td>
<td style="text-align:right;">
3148
</td>
<td style="text-align:right;">
1672
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:right;">
205
</td>
<td style="text-align:right;">
360
</td>
</tr>
<tr>
<td style="text-align:left;">
WT-2a_AAACGCTCAAACTCGT-1
</td>
<td style="text-align:left;">
WT-2a
</td>
<td style="text-align:right;">
2221
</td>
<td style="text-align:right;">
1236
</td>
<td style="text-align:right;">
2472
</td>
<td style="text-align:right;">
1378
</td>
<td style="text-align:right;">
0.0900495
</td>
<td style="text-align:right;">
0.1350743
</td>
<td style="text-align:right;">
0.2251238
</td>
<td style="text-align:right;">
2504
</td>
<td style="text-align:right;">
1235
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:right;">
142
</td>
<td style="text-align:right;">
251
</td>
</tr>
<tr>
<td style="text-align:left;">
WT-2a_AAACGCTCATGTTACG-1
</td>
<td style="text-align:left;">
WT-2a
</td>
<td style="text-align:right;">
2685
</td>
<td style="text-align:right;">
1518
</td>
<td style="text-align:right;">
2999
</td>
<td style="text-align:right;">
1694
</td>
<td style="text-align:right;">
0.0744879
</td>
<td style="text-align:right;">
0.8193669
</td>
<td style="text-align:right;">
0.8938547
</td>
<td style="text-align:right;">
2720
</td>
<td style="text-align:right;">
1509
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:right;">
176
</td>
<td style="text-align:right;">
314
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="calculate-per-sample-averages">Calculate per sample averages<a class="anchor" aria-label="anchor" href="#calculate-per-sample-averages"></a>
</h3>
<p>We can then use <code><a href="../reference/Median_Stats.html">Median_Stats()</a></code> to calculate per sample averages across all cells by supplying the new variables to the <code>median_var</code> parameter.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">median_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Median_Stats.html">Median_Stats</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">astrocytes_cortex</span>, group_by_var <span class="op">=</span> <span class="st">"orig.ident"</span>, median_var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCount_Diff"</span>,
    <span class="st">"nFeature_Diff"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
orig.ident
</th>
<th style="text-align:right;">
Median_nCount_RNA
</th>
<th style="text-align:right;">
Median_nFeature_RNA
</th>
<th style="text-align:right;">
Median_percent_mito
</th>
<th style="text-align:right;">
Median_percent_ribo
</th>
<th style="text-align:right;">
Median_percent_mito_ribo
</th>
<th style="text-align:right;">
Median_nCount_Diff
</th>
<th style="text-align:right;">
Median_nFeature_Diff
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
WT-2a
</td>
<td style="text-align:right;">
2949.5
</td>
<td style="text-align:right;">
1527.5
</td>
<td style="text-align:right;">
0.0224121
</td>
<td style="text-align:right;">
0.0673027
</td>
<td style="text-align:right;">
0.0948317
</td>
<td style="text-align:right;">
280
</td>
<td style="text-align:right;">
156
</td>
</tr>
<tr>
<td style="text-align:left;">
WT-6a
</td>
<td style="text-align:right;">
2765.0
</td>
<td style="text-align:right;">
1451.5
</td>
<td style="text-align:right;">
0.0352609
</td>
<td style="text-align:right;">
0.0656276
</td>
<td style="text-align:right;">
0.1065341
</td>
<td style="text-align:right;">
279
</td>
<td style="text-align:right;">
154
</td>
</tr>
<tr>
<td style="text-align:left;">
Totals (All Cells)
</td>
<td style="text-align:right;">
2875.0
</td>
<td style="text-align:right;">
1498.5
</td>
<td style="text-align:right;">
0.0293169
</td>
<td style="text-align:right;">
0.0665447
</td>
<td style="text-align:right;">
0.1006627
</td>
<td style="text-align:right;">
279
</td>
<td style="text-align:right;">
156
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="dual-assay-plotting">Dual Assay Plotting<a class="anchor" aria-label="anchor" href="#dual-assay-plotting"></a>
</h2>
<p>For Cell Bender especially, but also potentially for other assays as well, it can be helpful during analysis to plot the corrected and uncorrected counts for given feature. scCustomize contains function <code><a href="../reference/FeaturePlot_DualAssay.html">FeaturePlot_DualAssay()</a></code> to make easy.</p>
<p>Users just need to supply the names of the two assays to plot and the features. <em>NOTE: Make sure both assays have been normalized before plotting. The function will attempt to check and make sure both assays have been normalized but has not been tested in all scenarios.</em></p>
<div class="section level3">
<h3 id="example-plotting">Example Plotting<a class="anchor" aria-label="anchor" href="#example-plotting"></a>
</h3>
<p>For this example I’m using unpublished single nucleus RNA dataset from mouse cortex and have subsetted the astrocytes.</p>
<div class="section level4">
<h4 id="ambient-rna-gene">Ambient RNA gene<a class="anchor" aria-label="anchor" href="#ambient-rna-gene"></a>
</h4>
<p>First let’s plot gene that represents ambient RNA as it’s restricted in expression to neurons (synaptic gene). If Cell Bender has worked well we expect that expression of this gene will be very different between the two assays. <img src="Cell_Bender_Functions_files/figure-html/unnamed-chunk-11-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="non-ambient-rna-gene">Non-Ambient RNA gene<a class="anchor" aria-label="anchor" href="#non-ambient-rna-gene"></a>
</h4>
<p>Now let’s plot normally astrocyte restricted gene. If Cell Bender has worked well we expect that expression of this gene shouldn’t be very different between the two assays. <img src="Cell_Bender_Functions_files/figure-html/unnamed-chunk-12-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
