<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>General Helpers &amp; Utilities • scCustomize</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="General Helpers &amp; Utilities">
<meta property="og:description" content="scCustomize">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scCustomize</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Plotting Vignettes</li>
    <li>
      <a href="../articles/Gene_Expression_Plotting.html">Plotting #1: Analysis Plots</a>
    </li>
    <li>
      <a href="../articles/QC_Plots.html">Plotting #2: QC Plots</a>
    </li>
    <li>
      <a href="../articles/Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    </li>
    <li>
      <a href="../articles/Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Helpers &amp; Utilities</li>
    <li>
      <a href="../articles/Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    </li>
    <li>
      <a href="../articles/Markers_and_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    </li>
    <li>
      <a href="../articles/Statistics.html">Statistics Functions</a>
    </li>
    <li>
      <a href="../articles/Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    </li>
    <li>
      <a href="../articles/Misc_Functions.html">Misc Functions</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Package Interactivity/Functions</li>
    <li>
      <a href="../articles/LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    </li>
    <li>
      <a href="../articles/Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/samuel-marsh/scCustomize/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Helpers_and_Utilities_files/kePrint-0.0.1/kePrint.js"></script><link href="Helpers_and_Utilities_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>General Helpers &amp; Utilities</h1>
            
            <h4 data-toc-skip class="date">Compiled: December 16, 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/samuel-marsh/scCustomize/blob/HEAD/vignettes/Helpers_and_Utilities.Rmd" class="external-link"><code>vignettes/Helpers_and_Utilities.Rmd</code></a></small>
      <div class="hidden name"><code>Helpers_and_Utilities.Rmd</code></div>

    </div>

    
    
<hr>
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level1">
<h1 id="general-helperutility-functions">General Helper/Utility Functions<a class="anchor" aria-label="anchor" href="#general-helperutility-functions"></a>
</h1>
<p>scCustomize has several helper functions to simplify/streamline common tasks in scRNA-seq analysis. Let’s load packages and raw data object for this tutorial.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load Packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/samuel-marsh/scCustomize" class="external-link">scCustomize</a></span><span class="op">)</span>

<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">pbmc3k.SeuratData</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/pbmc3k.SeuratData/man/pbmc3k.html" class="external-link">pbmc3k</a></span></code></pre></div>
<div class="section level2">
<h2 id="add-mitochondrial-ribosomal-percentages">Add Mitochondrial &amp; Ribosomal Percentages<a class="anchor" aria-label="anchor" href="#add-mitochondrial-ribosomal-percentages"></a>
</h2>
<p>As discussed in <a href="https://samuel-marsh.github.io/scCustomize/articles/QC_Plots.html">QC Plotting vignette</a> one the first steps after creating object if often to calculate and add mitochondrial and ribosomal count percentages per cell/nucleus.</p>
<div class="section level3">
<h3 id="seurat-objects-add_mito_ribo_seurat">Seurat Objects: <code>Add_Mito_Ribo_Seurat()</code><a class="anchor" aria-label="anchor" href="#seurat-objects-add_mito_ribo_seurat"></a>
</h3>
<p>scCustomize contains easy wrapper function to automatically add both Mitochondrial and Ribosomal percentages to meta.data slot. If you are using mouse, human, or marmoset data all you need to do is specify the <code>species</code> parameter.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># These defaults can be run just by providing accepted species name</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></code></pre></div>
<p>To view list of accepted values for default species names simply set <code>list_species_names = TRUE</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat</a></span><span class="op">(</span>list_species_names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Mouse_Options
</th>
<th style="text-align:left;">
Human_Options
</th>
<th style="text-align:left;">
Marmoset_Options
</th>
<th style="text-align:left;">
Zebrafish_Options
</th>
<th style="text-align:left;">
Rat_Options
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
Mouse
</td>
<td style="text-align:left;">
Human
</td>
<td style="text-align:left;">
Marmoset
</td>
<td style="text-align:left;">
Zebrafish
</td>
<td style="text-align:left;">
Rat
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:left;">
mouse
</td>
<td style="text-align:left;">
human
</td>
<td style="text-align:left;">
marmoset
</td>
<td style="text-align:left;">
zebrafish
</td>
<td style="text-align:left;">
rat
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:left;">
Ms
</td>
<td style="text-align:left;">
Hu
</td>
<td style="text-align:left;">
CJ
</td>
<td style="text-align:left;">
DR
</td>
<td style="text-align:left;">
RN
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
ms
</td>
<td style="text-align:left;">
hu
</td>
<td style="text-align:left;">
Cj
</td>
<td style="text-align:left;">
Dr
</td>
<td style="text-align:left;">
Rn
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
Mm
</td>
<td style="text-align:left;">
Hs
</td>
<td style="text-align:left;">
cj
</td>
<td style="text-align:left;">
dr
</td>
<td style="text-align:left;">
rn
</td>
</tr>
<tr>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
mm
</td>
<td style="text-align:left;">
hs
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
</tbody>
</table>
<div class="section level4">
<h4 id="non-default-species">Non-default species<a class="anchor" aria-label="anchor" href="#non-default-species"></a>
</h4>
<p>However custom prefixes can be used for non-human/mouse/marmoset species with different annotations. Simply specify <code>species = other</code> and supply feature lists or regex patterns for your species of interest. <em>NOTE: If desired please submit issue on GitHub for additional default species. Please include regex pattern or list of genes for both mitochondrial and ribosomal genes and I will add additional built-in defaults to the function.</em></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Using gene name patterns</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"other"</span>, mito_pattern <span class="op">=</span> <span class="st">"regexp_pattern"</span>,
    ribo_pattern <span class="op">=</span> <span class="st">"regexp_pattern"</span><span class="op">)</span>

<span class="co"># Using feature name lists</span>
<span class="va">mito_gene_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"etc"</span><span class="op">)</span>
<span class="va">ribo_gene_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"etc"</span><span class="op">)</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"other"</span>, mito_features <span class="op">=</span> <span class="va">mito_gene_list</span>,
    ribo_features <span class="op">=</span> <span class="va">ribo_gene_list</span><span class="op">)</span>

<span class="co"># Using combination of gene lists and gene name patterns</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"Human"</span>, mito_features <span class="op">=</span> <span class="va">mito_gene_list</span>,
    ribo_pattern <span class="op">=</span> <span class="st">"regexp_pattern"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="warning-messages">Warning Messages<a class="anchor" aria-label="anchor" href="#warning-messages"></a>
</h3>
<p>The added benefit of <code>Add_Mito_Ribo_Seurat</code> &amp; <code>Add_Mito_Ribo_LIGER</code> is that they will return informative warnings if no Mitochondrial or Ribosomal features are found using the current species, features, or pattern specification.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># For demonstration purposes we can set `species = mouse` for this object of human cells</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Error in Add_Mito_Ribo_Seurat(seurat_object = pbmc, species = "mouse"): No Mito or Ribo features found in provided object using patterns/feature list provided.  Please check pattern/feature list and/or gene names in object.</span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Or if providing custom patterns/lists and features not found</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"other"</span>, mito_pattern <span class="op">=</span> <span class="st">"^MT-"</span>, ribo_pattern <span class="op">=</span> <span class="st">"BAD_PATTERN"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning in Add_Mito_Ribo_Seurat(seurat_object = pbmc, species = "other", : No</span>
<span class="co">## Ribo features found in provided object using pattern/feature list provided. No</span>
<span class="co">## column will be added to meta.data</span></code></pre>
<p><code>Add_Mito_Ribo_Seurat</code> and <code><a href="../reference/Add_Mito_Ribo_LIGER.html">Add_Mito_Ribo_LIGER()</a></code> will also return warnings if columns are already present in <code>@meta.data</code> slot and prompt you to provide override if you want to run the function.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Error in Add_Mito_Ribo_Seurat(seurat_object = pbmc, species = "human"): Columns with percent_mito and/or percent_ribo already present</span>
<span class="co">##   in meta.data slot.</span>
<span class="co">##   *To run function and overwrite columns set parameter `overwrite = TRUE` or change respective 'mito_name', 'ribo_name', or mito_ribo_name'*.</span></code></pre>
</div>
<div class="section level3">
<h3 id="liger-objects-add_mito_ribo_liger">LIGER Objects: <code>Add_Mito_Ribo_LIGER()</code><a class="anchor" aria-label="anchor" href="#liger-objects-add_mito_ribo_liger"></a>
</h3>
<p>scCustomize also contains identical function for use with LIGER objects. <code>Add_Mito_Ribo_LIGER</code> contains equivalent parameters and capabilities as <code>Add_Mito_Ribo_Seurat</code></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">liger_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">liger_obj</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="check-for-featuresgenes">Check for Features/Genes<a class="anchor" aria-label="anchor" href="#check-for-featuresgenes"></a>
</h2>
<p>scCustomize also makes forward-facing a number of utilities that are used internally in functions but may also have utility on their own.</p>
<div class="section level3">
<h3 id="gene_present-to-check-for-features-">
<code>Gene_Present()</code> to check for features.<a class="anchor" aria-label="anchor" href="#gene_present-to-check-for-features-"></a>
</h3>
<p><code>Gene_Present</code> is fairly basic function to check if feature exists in data. It can be used with Seurat or LIGER objects as well as generic data formats (Matrix, data.frame, tibble).</p>
<p>In addition to some warning messages <code>Gene_Present</code> returns a list with 3 entries when run:</p>
<ul>
<li>found_features: features found in the data.<br>
</li>
<li>bad_features: features not found in the data.<br>
</li>
<li>wrong_case_found_features: features found but in different case than present in input gene list.
<ul>
<li><em>If <code>bad_features</code> &gt; 0 then <code>Gene_Present</code> will convert <code>the gene list</code>bad_features` to all upper case and to sentence case and check against all possible features to see if wrong case was provided.</em></li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Example gene list with all examples (found genes, wrong case (lower) and misspelled (CD8A</span>
<span class="co"># forgetting to un-shift when typing 8))</span>
<span class="va">gene_input_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD14"</span>, <span class="st">"CD3E"</span>, <span class="st">"Cd4"</span>, <span class="st">"CD*A"</span><span class="op">)</span>

<span class="va">genes_present</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Gene_Present.html">Gene_Present</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">pbmc</span>, gene_list <span class="op">=</span> <span class="va">gene_input_list</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning in Gene_Present(data = pbmc, gene_list = gene_input_list): The following</span>
<span class="co">## features were omitted as they were not found: Cd4 and CD*A</span></code></pre>
<pre><code><span class="co">## Warning in Gene_Present(data = pbmc, gene_list = gene_input_list): NOTE: However, the following features were found: CD4.</span>
<span class="co">##         Please check intended case of features provided.</span></code></pre>
<p>Now let’s look at the output:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genes_present</span></code></pre></div>
<pre><code><span class="co">## $found_features</span>
<span class="co">## [1] "CD14" "CD3E"</span>
<span class="co">## </span>
<span class="co">## $bad_features</span>
<span class="co">## [1] "Cd4"  "CD*A"</span>
<span class="co">## </span>
<span class="co">## $wrong_case_found_features</span>
<span class="co">## [1] "CD4"</span></code></pre>
</div>
<div class="section level3">
<h3 id="turn-warningsmessages-off-">Turn warnings/messages off.<a class="anchor" aria-label="anchor" href="#turn-warningsmessages-off-"></a>
</h3>
<p>By default <code>Gene_Present</code> has 3 sets of warnings/messages it prints to console when it finds issues. If using the function yourself on its own or wrapped inside your own function and you prefer no messages each of these can be toggled using optional parameters.</p>
<ul>
<li>
<code>case_check_msg</code> prints and list of features if alternate case features are found in data.<br>
</li>
<li>
<code>omit_warn</code> prints warning and list of all features not found in data.<br>
</li>
<li>
<code>print_msg</code> prints message if all features in <code>gene_list</code> are found in data.</li>
</ul>
</div>
<div class="section level3">
<h3 id="check-for-updated-gene-symbols">Check for updated gene symbols<a class="anchor" aria-label="anchor" href="#check-for-updated-gene-symbols"></a>
</h3>
<p>In order to keep run times down and support offer greater support for offline use <code>Gene_Present</code> does not include a check for updated gene symbols. If you’re dataset is from human cells/donors you can simply supply the not found features from <code>Gene_Present</code> to Seurat’s <code>UpdateSymbolList</code> function.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_input_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD14"</span>, <span class="st">"CD3E"</span>, <span class="st">"Cd4"</span>, <span class="st">"CD*A"</span>, <span class="st">"SEPT1"</span><span class="op">)</span>

<span class="va">genes_present</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Gene_Present.html">Gene_Present</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">pbmc</span>, gene_list <span class="op">=</span> <span class="va">gene_input_list</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning in Gene_Present(data = pbmc, gene_list = gene_input_list): The following</span>
<span class="co">## features were omitted as they were not found: Cd4 and CD*A</span></code></pre>
<pre><code><span class="co">## Warning in Gene_Present(data = pbmc, gene_list = gene_input_list): NOTE: However, the following features were found: CD4.</span>
<span class="co">##         Please check intended case of features provided.</span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">check_symbols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/UpdateSymbolList.html" class="external-link">UpdateSymbolList</a></span><span class="op">(</span>symbols <span class="op">=</span> <span class="va">genes_present</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: No updated symbols found</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="merging-dataobjects">Merging Data/Objects<a class="anchor" aria-label="anchor" href="#merging-dataobjects"></a>
</h2>
<div class="section level3">
<h3 id="merging-raw-data">Merging raw data<a class="anchor" aria-label="anchor" href="#merging-raw-data"></a>
</h3>
<p>It can often be advantageous to merge raw data before creating analysis objects vs creating lots of objects and merging them all later. scCustomize features a modified version of the internal LIGER function <code>MergeSparseDataAll()</code>.</p>
<p><code><a href="../reference/Merge_Sparse_Data_All.html">Merge_Sparse_Data_All()</a></code> will combine a list of sparse matrices and return single sparse matrix. Additionally, by specifying the <code>add_cell_ids</code> parameter you can specify a prefix to be added to the barcodes from each entry in the list (using “_” as delimiter).</p>
<p>This function can be especially useful when combined with any of the scCustomize’s <code>Read_</code> data functions which automatically return named lists of matrices and the ability to specify sample <code>orig.ident</code> when creating Seurat objects. See <a href="https://samuel-marsh.github.io/scCustomize/articles/Read_and_Write_Functions.html">Read &amp; Write Vignette</a> for more info on the data import functions.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Read in data</span>
<span class="va">GEO_10X</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Read10X_GEO.html">Read10X_GEO</a></span><span class="op">(</span>data_dir <span class="op">=</span> <span class="st">"assets/GSE152183_RAW_Marsh/"</span><span class="op">)</span>

<span class="co"># Merge data and add sample prefix</span>
<span class="va">GEO_10X_merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Merge_Sparse_Data_All.html">Merge_Sparse_Data_All</a></span><span class="op">(</span>matrix_list <span class="op">=</span> <span class="va">GEO_10X</span>, add_cell_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">GEO_10X</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Create Seurat Object and specify orig.ident location</span>
<span class="va">GEO_10X_Seurat</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">GEO_10X_merged</span>, names.field <span class="op">=</span> <span class="fl">1</span>, names.delim <span class="op">=</span> <span class="st">"_"</span>,
    min.features <span class="op">=</span> <span class="fl">200</span>, min.cells <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="merging-list-of-seurat-objects">Merging List of Seurat Objects<a class="anchor" aria-label="anchor" href="#merging-list-of-seurat-objects"></a>
</h3>
<p>Sometimes it can be advantageous to create a list of multiple Seurat Objects in order to run similar pipeline on all objects in loop.</p>
<p>To facilitate ease in merging such lists into single object scCustomize contains simple wrapper <code>Merge_Seurat_List</code> that uses <code><a href="https://purrr.tidyverse.org/reference/reduce.html" class="external-link">purrr::reduce()</a></code> to merge all objects in list into single combined object</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">list_of_objects</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">obj1</span>, <span class="va">obj2</span>, <span class="va">obj2</span>, <span class="va">...</span>, <span class="va">obj10</span><span class="op">)</span>

<span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Merge_Seurat_List.html">Merge_Seurat_List</a></span><span class="op">(</span>list_seurat <span class="op">=</span> <span class="va">list_of_objects</span><span class="op">)</span>

<span class="co"># Can also add sample specific ids to each object during the merge</span>
<span class="va">cell_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span>, <span class="st">"sample3"</span>, <span class="va">...</span>, <span class="st">"sample10"</span><span class="op">)</span>

<span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Merge_Seurat_List.html">Merge_Seurat_List</a></span><span class="op">(</span>list_seurat <span class="op">=</span> <span class="va">list_of_objects</span>, add.cell.ids <span class="op">=</span> <span class="va">cell_ids</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="storing-misc-information-in-seurat-objects">Storing Misc Information in Seurat Objects<a class="anchor" aria-label="anchor" href="#storing-misc-information-in-seurat-objects"></a>
</h2>
<p>Seurat objects contain an extra empty slot that can be used to store any extra information desired.<br>
scCustomize contains two functions <code>Store_Misc_Info_Seurat</code> and a wrapper around that function <code>Store_Palette_Seurat</code> to make this process easy.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Data can be vectors or data.frames</span>
<span class="va">misc_info</span> <span class="op">&lt;-</span> <span class="st">"misc_vector_dataframe_list_etc"</span>

<span class="co"># Add data to the @misc slot in Seurat Object</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Store_Misc_Info_Seurat.html">Store_Misc_Info_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, data_to_store <span class="op">=</span> <span class="va">misc_info</span>, data_name <span class="op">=</span> <span class="st">"misc_info_name"</span><span class="op">)</span></code></pre></div>
<div class="section level3">
<h3 id="storing-lists">Storing Lists<a class="anchor" aria-label="anchor" href="#storing-lists"></a>
</h3>
<p>If you are storing a list in the <code>@misc</code> slot there is additional parameter that dictates whether to store the information as a list or whether to store each entry in the list separately.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create list</span>
<span class="va">misc_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"misc_item1"</span>, <span class="st">"misc_item2"</span>, <span class="va">etc</span><span class="op">)</span>

<span class="co"># Store the list directly</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Store_Misc_Info_Seurat.html">Store_Misc_Info_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, data_to_store <span class="op">=</span> <span class="va">misc_info</span>, data_name <span class="op">=</span> <span class="st">"misc_info_name"</span>,
    list_as_list <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># Store each entry in list as separate entity in `@misc` slot</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Store_Misc_Info_Seurat.html">Store_Misc_Info_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, data_to_store <span class="op">=</span> <span class="va">misc_info</span>, data_name <span class="op">=</span> <span class="st">"misc_info_name"</span>,
    list_as_list <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="storing-color-palettes">Storing Color Palettes<a class="anchor" aria-label="anchor" href="#storing-color-palettes"></a>
</h3>
<p>One of the most common times I use this function is to store color palettes associated with clustering or subclustering.<br>
To make it easier to remember function call in this situation scCustomize contains a wrapper function <code>Store_Palette_Seurat</code>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Data can be vectors or data.frames</span>
<span class="va">annotated_color_palette</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"color1"</span>, <span class="st">"color2"</span>, <span class="st">"color3"</span>, <span class="st">"etc"</span><span class="op">)</span>

<span class="co"># Add data to the @misc slot in Seurat Object</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Store_Palette_Seurat.html">Store_Palette_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, palette <span class="op">=</span> <span class="va">annotated_color_palette</span>, palette_name <span class="op">=</span> <span class="st">"Round01_Color_Pal"</span><span class="op">)</span>

<span class="co"># Then you can easily call that palette (with tab completion) when plotting without ever</span>
<span class="co"># needing to reload the palette in current environment</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, cols <span class="op">=</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">Round01_Color_Pal</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="modify-cell-barcodes-in-raw-data">Modify Cell Barcodes in Raw Data<a class="anchor" aria-label="anchor" href="#modify-cell-barcodes-in-raw-data"></a>
</h2>
<p>Sometimes, especially with public data, you may want to modify the cell barcode names before creating analysis object.</p>
<p>scCustomize contains a selection of functions to simplify this process:</p>
<div class="section level3">
<h3 id="replace-suffixes">Replace Suffixes<a class="anchor" aria-label="anchor" href="#replace-suffixes"></a>
</h3>
<p><code>Replace_Suffix</code> can be used on single matrix/data.frame or list of matrices/data.frames to modify to remove suffixes</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># For single object</span>
<span class="va">data_mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Replace_Suffix.html">Replace_Suffix</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">raw_data</span>, current_suffix <span class="op">=</span> <span class="st">"-1"</span>, new_suffix <span class="op">=</span> <span class="st">"-2"</span><span class="op">)</span>

<span class="co"># For list of objects containing same suffix</span>
<span class="va">raw_data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">raw_data1</span>, <span class="va">raw_data2</span>, <span class="va">raw_data3</span>, <span class="va">...</span>, <span class="va">raw_data10</span><span class="op">)</span>

<span class="va">new_suffixes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-1"</span>, <span class="st">"-2"</span>, <span class="st">"-3"</span>, <span class="va">...</span>, <span class="st">"-10"</span><span class="op">)</span>

<span class="va">data_mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Replace_Suffix.html">Replace_Suffix</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">raw_data_list</span>, current_suffix <span class="op">=</span> <span class="st">"-1"</span>, new_suffix <span class="op">=</span> <span class="va">new_suffixes</span><span class="op">)</span>


<span class="co"># For list of objects containing different suffixes</span>
<span class="va">raw_data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">raw_data1</span>, <span class="va">raw_data2</span>, <span class="va">raw_data3</span>, <span class="va">...</span>, <span class="va">raw_data10</span><span class="op">)</span>

<span class="va">old_suffixes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-A"</span>, <span class="st">"-B"</span>, <span class="st">"-C"</span>, <span class="va">...</span>, <span class="st">"-J"</span><span class="op">)</span>
<span class="va">new_suffixes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-1"</span>, <span class="st">"-2"</span>, <span class="st">"-3"</span>, <span class="va">...</span>, <span class="st">"-10"</span><span class="op">)</span>

<span class="va">data_mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Replace_Suffix.html">Replace_Suffix</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">raw_data_list</span>, current_suffix <span class="op">=</span> <span class="va">old_suffixes</span>, new_suffix <span class="op">=</span> <span class="va">new_suffixes</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="strip-suffixes">Strip Suffixes<a class="anchor" aria-label="anchor" href="#strip-suffixes"></a>
</h3>
<p><code>Replace_Suffix</code> can also be used to strip suffixes from data</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># For single object</span>
<span class="va">data_mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Replace_Suffix.html">Replace_Suffix</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">raw_data</span>, current_suffix <span class="op">=</span> <span class="st">"-1"</span>, new_suffix <span class="op">=</span> <span class="st">""</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="change-prefixsuffix-delimiters">Change Prefix/Suffix Delimiters<a class="anchor" aria-label="anchor" href="#change-prefixsuffix-delimiters"></a>
</h3>
<p>scCustomize has 3 functions to facilitate changing the type of delimiters present in cell barcodes.</p>
<ul>
<li>
<code><a href="../reference/Change_Delim_Prefix.html">Change_Delim_Prefix()</a></code> Change just the suffix delimiter.</li>
<li>
<code><a href="../reference/Change_Delim_Suffix.html">Change_Delim_Suffix()</a></code> Change just the prefix delimiter.</li>
<li>
<code><a href="../reference/Change_Delim_All.html">Change_Delim_All()</a></code> Change all delimiters.</li>
</ul>
<p>These functions all take identical inputs and can be applied to either single matrix/data.frames or lists of matrices/data.frames.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data_mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Change_Delim_Prefix.html">Change_Delim_Prefix</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">raw_data</span>, current_delim <span class="op">=</span> <span class="st">"."</span>, new_delim <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>

<span class="va">data_mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Change_Delim_Suffix.html">Change_Delim_Suffix</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">raw_data</span>, current_delim <span class="op">=</span> <span class="st">"."</span>, new_delim <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>

<span class="va">data_mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Change_Delim_All.html">Change_Delim_All</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">raw_data</span>, current_delim <span class="op">=</span> <span class="st">"."</span>, new_delim <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></code></pre></div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
