<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scCustomize">
<title>Plotting #2: QC Plots • scCustomize</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plotting #2: QC Plots">
<meta property="og:description" content="scCustomize">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scCustomize</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.1.0004</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">Getting Started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/Installation.html">Install</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-vignettes">Vignettes</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-vignettes">
    <h6 class="dropdown-header" data-toc-skip>Plotting Vignettes</h6>
    <a class="dropdown-item" href="../articles/Gene_Expression_Plotting.html">Plotting #1: Analysis Plots</a>
    <a class="dropdown-item" href="../articles/QC_Plots.html">Plotting #2: QC Plots</a>
    <a class="dropdown-item" href="../articles/Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    <a class="dropdown-item" href="../articles/Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    <a class="dropdown-item" href="../articles/Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Helpers &amp; Utilities</h6>
    <a class="dropdown-item" href="../articles/Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    <a class="dropdown-item" href="../articles/Markers_and_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    <a class="dropdown-item" href="../articles/Statistics.html">Statistics Functions</a>
    <a class="dropdown-item" href="../articles/Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    <a class="dropdown-item" href="../articles/Misc_Functions.html">Misc Functions</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Other Package Interactivity/Functions</h6>
    <a class="dropdown-item" href="../articles/LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    <a class="dropdown-item" href="../articles/Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News/Changelog</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/samuel-marsh/scCustomize/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">
<script src="QC_Plots_files/kePrint-0.0.1/kePrint.js"></script><link href="QC_Plots_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Plotting #2: QC Plots</h1>
            
            <h4 data-toc-skip class="date">Compiled: December 01, 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/samuel-marsh/scCustomize/blob/HEAD/vignettes/QC_Plots.Rmd" class="external-link"><code>vignettes/QC_Plots.Rmd</code></a></small>
      <div class="d-none name"><code>QC_Plots.Rmd</code></div>
    </div>

    
    
<hr>
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level1">
<h1 id="qc-plots">QC Plots<a class="anchor" aria-label="anchor" href="#qc-plots"></a>
</h1>
<p>One of the first steps in all scRNA-seq analyses is performing a number of QC checks and plots so that data can be appropriately filtered. scCustomize contains a number of functions that can be used to quickly and easily generate some of the most relevant QC plots.</p>
<p>For this tutorial, I will be utilizing HCA bone marrow cell data from the SeuratData package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/samuel-marsh/scCustomize" class="external-link">scCustomize</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/traversc/qs" class="external-link">qs</a></span><span class="op">)</span>

<span class="co"># Load Example Dataset</span>
<span class="va">hca_bm</span> <span class="op">&lt;-</span> <span class="fu">hcabm40k.SeuratData</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/hcabm40k.SeuratData/man/hcabm40k.html" class="external-link">hcabm40k</a></span>

<span class="co"># Add pseudo group variable just for this vignette</span>
<span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">group</span><span class="op">[</span><span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"MantonBM1"</span> <span class="op">|</span> <span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span>
    <span class="st">"MantonBM2"</span> <span class="op">|</span> <span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"MantonBM3"</span> <span class="op">|</span> <span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"MantonBM4"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Group 1"</span>

<span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">group</span><span class="op">[</span><span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"MantonBM5"</span> <span class="op">|</span> <span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span>
    <span class="st">"MantonBM6"</span> <span class="op">|</span> <span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"MantonBM7"</span> <span class="op">|</span> <span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"MantonBM8"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Group 2"</span></code></pre></div>
<div class="section level2">
<h2 id="add-mitochondrial-and-ribosomal-gene-percentages">Add Mitochondrial and Ribosomal Gene Percentages<a class="anchor" aria-label="anchor" href="#add-mitochondrial-and-ribosomal-gene-percentages"></a>
</h2>
<p>scCustomize contains easy wrapper function to automatically add both Mitochondrial and Ribosomal count percentages to meta.data slot. If you are using mouse, human, or marmoset data all you need to do is specify the <code>species</code> parameter.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># These defaults can be run just by providing accepted species name</span>
<span class="va">hca_bm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, species <span class="op">=</span> <span class="st">"Human"</span><span class="op">)</span></code></pre></div>
<p><em>NOTE: There is analogous function for LIGER objects (see: <code>Add_Mito_Ribo_LIGER</code>) but QC plotting functions currently only support Seurat objects.</em></p>
To view list of accepted values for default species names simply set <code>list_species_names = TRUE</code>.
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
Mouse_Options
</th>
<th style="text-align:left;">
Human_Options
</th>
<th style="text-align:left;">
Marmoset_Options
</th>
<th style="text-align:left;">
Zebrafish_Options
</th>
<th style="text-align:left;">
Rat_Options
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
Mouse
</td>
<td style="text-align:left;">
Human
</td>
<td style="text-align:left;">
Marmoset
</td>
<td style="text-align:left;">
Zebrafish
</td>
<td style="text-align:left;">
Rat
</td>
</tr>
<tr>
<td style="text-align:left;">
mouse
</td>
<td style="text-align:left;">
human
</td>
<td style="text-align:left;">
marmoset
</td>
<td style="text-align:left;">
zebrafish
</td>
<td style="text-align:left;">
rat
</td>
</tr>
<tr>
<td style="text-align:left;">
Ms
</td>
<td style="text-align:left;">
Hu
</td>
<td style="text-align:left;">
CJ
</td>
<td style="text-align:left;">
DR
</td>
<td style="text-align:left;">
RN
</td>
</tr>
<tr>
<td style="text-align:left;">
ms
</td>
<td style="text-align:left;">
hu
</td>
<td style="text-align:left;">
Cj
</td>
<td style="text-align:left;">
Dr
</td>
<td style="text-align:left;">
Rn
</td>
</tr>
<tr>
<td style="text-align:left;">
Mm
</td>
<td style="text-align:left;">
Hs
</td>
<td style="text-align:left;">
cj
</td>
<td style="text-align:left;">
dr
</td>
<td style="text-align:left;">
rn
</td>
</tr>
<tr>
<td style="text-align:left;">
mm
</td>
<td style="text-align:left;">
hs
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
</tbody>
</table>
<p>However custom prefixes can be used for non-human/mouse/marmoset species with different annotations. Simply specify <code>species = other</code> and supply feature lists or regex patterns for your species of interest.<br><em>NOTE: If desired please submit issue on GitHub for additional default species. Please include regex pattern or list of genes for both mitochondrial and ribosomal genes and I will add additional built-in defaults to the function.</em></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Using gene name patterns</span>
<span class="va">hca_bm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, species <span class="op">=</span> <span class="st">"other"</span>, mito_pattern <span class="op">=</span> <span class="st">"regexp_pattern"</span>,
    ribo_pattern <span class="op">=</span> <span class="st">"regexp_pattern"</span><span class="op">)</span>

<span class="co"># Using feature name lists</span>
<span class="va">mito_gene_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"etc"</span><span class="op">)</span>
<span class="va">ribo_gene_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"etc"</span><span class="op">)</span>
<span class="va">hca_bm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, species <span class="op">=</span> <span class="st">"other"</span>, mito_features <span class="op">=</span> <span class="va">mito_gene_list</span>,
    ribo_features <span class="op">=</span> <span class="va">ribo_gene_list</span><span class="op">)</span>

<span class="co"># Using combination of gene lists and gene name patterns</span>
<span class="va">hca_bm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, species <span class="op">=</span> <span class="st">"Human"</span>, mito_features <span class="op">=</span> <span class="va">mito_gene_list</span>,
    ribo_pattern <span class="op">=</span> <span class="st">"regexp_pattern"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="plotting-qc-metrics">Plotting QC Metrics<a class="anchor" aria-label="anchor" href="#plotting-qc-metrics"></a>
</h2>
<p>scCustomize has a number of quick QC plotting options for ease of use.<br><em>NOTE: Most scCustomize plotting functions contain <code>...</code> parameter to allow user to supply any of the parameters for the original Seurat function that is being used under the hood.</em></p>
</div>
<div class="section level2">
<h2 id="vlnplot-based-qc-plots">
<code>VlnPlot</code>-Based QC Plots<a class="anchor" aria-label="anchor" href="#vlnplot-based-qc-plots"></a>
</h2>
<p>scCustomize contains 4 functions which wrap <code><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">Seurat::VlnPlot()</a></code>.</p>
<ul>
<li>
<code>QC_Plots_Genes</code> Plots genes per cell/nucleus.</li>
<li>
<code>QC_Plots_UMIs</code> Plots UMIs per cell/nucleus.<br>
</li>
<li>
<code>QC_Plots_Mito</code> Plots mito% (named “percent_mito”) per cell/nucleus.</li>
<li>
<code>QC_Plots_Feature</code> Plots “feature” per cell/nucleus. Using parameter <code>feature</code> to allow plotting of any applicable named feature in object@meta.data slot.</li>
</ul>
<p>scCustomize functions have the added benefit of:</p>
<ul>
<li>Feature to plot set by default (except for <code>QC_Plots_Feature</code>).</li>
<li>Added high/low cutoff parameters to allow for easy visualization of potential cutoff thresholds.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># All functions contain</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_Genes.html">QC_Plots_Genes</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff <span class="op">=</span> <span class="fl">800</span>, high_cutoff <span class="op">=</span> <span class="fl">5500</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_UMIs.html">QC_Plots_UMIs</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff <span class="op">=</span> <span class="fl">1200</span>, high_cutoff <span class="op">=</span> <span class="fl">45000</span><span class="op">)</span>
<span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_Mito.html">QC_Plots_Mito</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, high_cutoff <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="QC_Plots_files/figure-html/unnamed-chunk-7-1.png" width="1248"></p>
<div class="section level3">
<h3 id="additional-parameters">Additional parameters<a class="anchor" aria-label="anchor" href="#additional-parameters"></a>
</h3>
<p>In addition to being able to supply Seurat parameters with <code>...</code> these plots like many others in scCustomize contain other additional parameters to customize plot output without need for post-plot ggplot2 modifications</p>
<ul>
<li>
<code>plot_title</code>: Change plot title</li>
<li>
<code>x_axis_label</code>/<code>y_axis_label</code>: Change axis labels.</li>
<li>
<code>x_lab_rotate</code>: Should x-axis label be rotated 45 degrees?</li>
<li>
<code>y_axis_log</code>: Should y-axis in linear or log10 scale.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_UMIs.html">QC_Plots_UMIs</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff <span class="op">=</span> <span class="fl">1200</span>, high_cutoff <span class="op">=</span> <span class="fl">45000</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_UMIs.html">QC_Plots_UMIs</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff <span class="op">=</span> <span class="fl">1200</span>, high_cutoff <span class="op">=</span> <span class="fl">45000</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span>,
    y_axis_log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="QC_Plots_files/figure-html/unnamed-chunk-8-1.png" alt="*Setting `y_axis_log` can be very helpful for initial plots where outliers skew the visualization of the majority of the data without excluding data by setting y-axis limit.*" width="1248"><p class="caption">
<em>Setting <code>y_axis_log</code> can be very helpful for initial plots where outliers skew the visualization of the majority of the data without excluding data by setting y-axis limit.</em>
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="featurescatter-based-qc-plots">
<code>FeatureScatter</code>-Based QC Plots<a class="anchor" aria-label="anchor" href="#featurescatter-based-qc-plots"></a>
</h2>
<p>scCustomize contains 3 functions which wrap <code><a href="https://satijalab.org/seurat/reference/FeatureScatter.html" class="external-link">Seurat::FeatureScatter()</a></code> with added visualization of potential cutoff thresholds and some additional functionality:</p>
<ul>
<li>
<code>QC_Plot_UMIvsGene</code> Plots genes vs UMIs per cell/nucleus</li>
<li>
<code>QC_Plot_GenevsFeature</code> Plots Genes vs. “feature” per cell/nucleus. Using parameter <code>feature1</code> to allow plotting of any applicable named feature in object@meta.data slot.<br>
</li>
<li>
<code>QC_Plot_UMIvsFeature</code> Plots UMIs vs. “feature” per cell/nucleus. Using parameter <code>feature1</code> to allow plotting of any applicable named feature in object@meta.data slot.</li>
</ul>
<div class="section level3">
<h3 id="newmodified-functionality">New/Modified functionality<a class="anchor" aria-label="anchor" href="#newmodified-functionality"></a>
</h3>
<ul>
<li>Better default color palettes</li>
<li>
<code>shuffle = TRUE</code> by default to prevent hiding of datasets</li>
<li>Ability to set &amp; visualize potential cutoff thresholds (similar to VlnPlot based QC Plots above)</li>
<li>Report potential post filtering correlation in addition to whole dataset correlation when using <code>QC_Plot_UMIvsGene</code> (based on values provided to high and low cutoff parameters)</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># All functions contain</span>
<span class="fu"><a href="../reference/QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff_gene <span class="op">=</span> <span class="fl">800</span>, high_cutoff_gene <span class="op">=</span> <span class="fl">5500</span>, low_cutoff_UMI <span class="op">=</span> <span class="fl">500</span>,
    high_cutoff_UMI <span class="op">=</span> <span class="fl">50000</span><span class="op">)</span>
<span class="fu"><a href="../reference/QC_Plot_GenevsFeature.html">QC_Plot_GenevsFeature</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, feature1 <span class="op">=</span> <span class="st">"percent_mito"</span>, low_cutoff_gene <span class="op">=</span> <span class="fl">800</span>,
    high_cutoff_gene <span class="op">=</span> <span class="fl">5500</span>, high_cutoff_feature <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<p><img src="QC_Plots_files/figure-html/unnamed-chunk-10-1.png" width="1248" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="color-data-by-continuous-meta-data-variable">Color data by continuous meta data variable<a class="anchor" aria-label="anchor" href="#color-data-by-continuous-meta-data-variable"></a>
</h3>
<p><code>QC_Plot_UMIvsGene</code> contains the ability to color points by continuous meta data variables.<br>
This can be used to plot % of mito reads in addition to UMI vs. Gene comparisons</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, meta_gradient_name <span class="op">=</span> <span class="st">"percent_mito"</span>, low_cutoff_gene <span class="op">=</span> <span class="fl">800</span>,
    high_cutoff_gene <span class="op">=</span> <span class="fl">5500</span>, high_cutoff_UMI <span class="op">=</span> <span class="fl">45000</span><span class="op">)</span>
<span class="fu"><a href="../reference/QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, meta_gradient_name <span class="op">=</span> <span class="st">"percent_mito"</span>, low_cutoff_gene <span class="op">=</span> <span class="fl">800</span>,
    high_cutoff_gene <span class="op">=</span> <span class="fl">5500</span>, high_cutoff_UMI <span class="op">=</span> <span class="fl">45000</span>, meta_gradient_low_cutoff <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="QC_Plots_files/figure-html/unnamed-chunk-12-1.png" alt="*`QC_Plot_UMIvsGene()` when using `meta_gradient_name` outputs plot colored by meta data variable (left) to view only points above potential cutoff `meta_gradient_low_cutoff` can be specified to alter the plotting (right).*" width="1248"><p class="caption">
<em><code><a href="../reference/QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene()</a></code> when using <code>meta_gradient_name</code> outputs plot colored by meta data variable (left) to view only points above potential cutoff <code>meta_gradient_low_cutoff</code> can be specified to alter the plotting (right).</em>
</p>
</div>
</div>
<div class="section level3">
<h3 id="combination-plots">Combination Plots<a class="anchor" aria-label="anchor" href="#combination-plots"></a>
</h3>
<p>If you are interested in viewing <code>QC_Plot_UMIvsGene</code> both by discrete grouping variable and by continuous variable without writing function twice you can use <code>combination = TRUE</code> and plot output will contain both plots.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, meta_gradient_name <span class="op">=</span> <span class="st">"percent_mito"</span>, low_cutoff_gene <span class="op">=</span> <span class="fl">800</span>,
    high_cutoff_gene <span class="op">=</span> <span class="fl">5500</span>, high_cutoff_UMI <span class="op">=</span> <span class="fl">45000</span>, meta_gradient_low_cutoff <span class="op">=</span> <span class="fl">20</span>, combination <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="QC_Plots_files/figure-html/unnamed-chunk-13-1.png" alt="*`QC_Plot_UMIvsGene()` when using `combination = TRUE` will output both the Gene x UMI by active identity and with meta data gradient coloring.*" width="1248"><p class="caption">
<em><code><a href="../reference/QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene()</a></code> when using <code>combination = TRUE</code> will output both the Gene x UMI by active identity and with meta data gradient coloring.</em>
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="analyze-median-qc-values-per-samplelibrary">Analyze Median QC Values per Sample/Library<a class="anchor" aria-label="anchor" href="#analyze-median-qc-values-per-samplelibrary"></a>
</h2>
<p>scCustomize also contains a few helpful functions for returning and plotting the median values for these metrics on per sample/library basis.</p>
<div class="section level3">
<h3 id="calculate-median-values-return-data-frame">Calculate Median Values &amp; Return data.frame<a class="anchor" aria-label="anchor" href="#calculate-median-values-return-data-frame"></a>
</h3>
<p>scCustomize contains function <code>Median_Stats</code> to quickly calculate the medians for basic QC stats (Genes/, UMIs/, %Mito/Cell, etc) and return a data.frame.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">median_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Median_Stats.html">Median_Stats</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, group_by_var <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
orig.ident
</th>
<th style="text-align:right;">
Median_nCount_RNA
</th>
<th style="text-align:right;">
Median_nFeature_RNA
</th>
<th style="text-align:right;">
Median_percent_mito
</th>
<th style="text-align:right;">
Median_percent_ribo
</th>
<th style="text-align:right;">
Median_percent_mito_ribo
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
MantonBM1
</td>
<td style="text-align:right;">
2634.5
</td>
<td style="text-align:right;">
807
</td>
<td style="text-align:right;">
3.928074
</td>
<td style="text-align:right;">
42.04921
</td>
<td style="text-align:right;">
46.43790
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM2
</td>
<td style="text-align:right;">
2462.5
</td>
<td style="text-align:right;">
783
</td>
<td style="text-align:right;">
3.556994
</td>
<td style="text-align:right;">
42.96184
</td>
<td style="text-align:right;">
46.87269
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM3
</td>
<td style="text-align:right;">
2617.0
</td>
<td style="text-align:right;">
769
</td>
<td style="text-align:right;">
3.301410
</td>
<td style="text-align:right;">
40.65558
</td>
<td style="text-align:right;">
44.10722
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM4
</td>
<td style="text-align:right;">
2264.0
</td>
<td style="text-align:right;">
780
</td>
<td style="text-align:right;">
3.384891
</td>
<td style="text-align:right;">
33.73245
</td>
<td style="text-align:right;">
37.96850
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM5
</td>
<td style="text-align:right;">
2677.0
</td>
<td style="text-align:right;">
741
</td>
<td style="text-align:right;">
3.265677
</td>
<td style="text-align:right;">
41.55203
</td>
<td style="text-align:right;">
45.33713
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM6
</td>
<td style="text-align:right;">
2677.0
</td>
<td style="text-align:right;">
835
</td>
<td style="text-align:right;">
3.202053
</td>
<td style="text-align:right;">
43.07874
</td>
<td style="text-align:right;">
46.70119
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM7
</td>
<td style="text-align:right;">
2429.0
</td>
<td style="text-align:right;">
732
</td>
<td style="text-align:right;">
3.216184
</td>
<td style="text-align:right;">
44.68831
</td>
<td style="text-align:right;">
48.20765
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM8
</td>
<td style="text-align:right;">
2453.0
</td>
<td style="text-align:right;">
701
</td>
<td style="text-align:right;">
3.697331
</td>
<td style="text-align:right;">
44.97911
</td>
<td style="text-align:right;">
48.96420
</td>
</tr>
<tr>
<td style="text-align:left;">
Totals (All Cells)
</td>
<td style="text-align:right;">
2524.0
</td>
<td style="text-align:right;">
769
</td>
<td style="text-align:right;">
3.447040
</td>
<td style="text-align:right;">
41.53183
</td>
<td style="text-align:right;">
45.40989
</td>
</tr>
</tbody>
</table>
<p>The <code>Median_Stats</code> function has some column names stored by default but will also calculate medians for additional meta.data columns using the optional <code>median_var</code> parameter</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">median_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Median_Stats.html">Median_Stats</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, group_by_var <span class="op">=</span> <span class="st">"orig.ident"</span>, median_var <span class="op">=</span> <span class="st">"meta_data_column_name"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plotting-median-values">Plotting Median Values<a class="anchor" aria-label="anchor" href="#plotting-median-values"></a>
</h3>
<p>scCustomize also contains a few functions to plot some of these median value calculations, which can be used on their own without need to return data.frame first.</p>
<ul>
<li>
<code>Plot_Median_Genes</code><br>
</li>
<li>
<code>Plot_Median_UMIs</code><br>
</li>
<li>
<code>Plot_Median_Mito</code><br>
</li>
<li>
<code>Plot_Median_Other</code>
<ul>
<li><em>Used to plot any other numeric variable present in object meta.data slot.</em></li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/Plot_Median_Genes.html">Plot_Median_Genes</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, group_by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span>
<span class="fu"><a href="../reference/Plot_Median_UMIs.html">Plot_Median_UMIs</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, group_by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span>
<span class="fu"><a href="../reference/Plot_Median_Mito.html">Plot_Median_Mito</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, group_by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span>
<span class="fu"><a href="../reference/Plot_Median_Other.html">Plot_Median_Other</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, median_var <span class="op">=</span> <span class="st">"percent_ribo"</span>, group_by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></code></pre></div>
<p><img src="QC_Plots_files/figure-html/unnamed-chunk-18-1.png" width="1248" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="plot-number-of-cellsnuclei-per-sample">Plot Number of Cells/Nuclei per Sample<a class="anchor" aria-label="anchor" href="#plot-number-of-cellsnuclei-per-sample"></a>
</h3>
<p>scCustomize also contains plotting function to plot the number of cells or nuclei per sample.</p>
<p>Since the HCA Bone Marrow dataset has exactly the same number of cells per sample we will use the microglia object from the Analysis Plots vignette.</p>
<p><img src="QC_Plots_files/figure-html/unnamed-chunk-20-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.0.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
