<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ scCustomize</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scCustomize</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">Getting Started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/Installation.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-vignettes">Vignettes</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-vignettes">
    <h6 class="dropdown-header" data-toc-skip>Plotting Vignettes</h6>
    <a class="dropdown-item" href="../articles/Gene_Expression_Plotting.html">Plotting #1: Analysis Plots</a>
    <a class="dropdown-item" href="../articles/QC_Plots.html">Plotting #2: QC Plots</a>
    <a class="dropdown-item" href="../articles/Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    <a class="dropdown-item" href="../articles/Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    <a class="dropdown-item" href="../articles/Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Helpers &amp; Utilities</h6>
    <a class="dropdown-item" href="../articles/Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    <a class="dropdown-item" href="../articles/Markers_and_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    <a class="dropdown-item" href="../articles/Statistics.html">Statistics Functions</a>
    <a class="dropdown-item" href="../articles/Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    <a class="dropdown-item" href="../articles/Misc_Functions.html">Misc Functions</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Other Package Interactivity/Functions</h6>
    <a class="dropdown-item" href="../articles/LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    <a class="dropdown-item" href="../articles/Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News/Changelog</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/FAQ.html">FAQ</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/samuel-marsh/scCustomize/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="read-write-data">Read &amp; Write Data<a class="anchor" aria-label="anchor" href="#read-write-data"></a></h2>
      
      <p class="section-desc"></p><p>Functions for reading and writing single cell sequencing data from raw data files.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="Create_10X_H5.html">Create_10X_H5()</a></code> 
        </dt>
        <dd>Create H5 from 10X Outputs</dd>
      </dl><dl><dt>
          
          <code><a href="Create_CellBender_Merged_Seurat.html">Create_CellBender_Merged_Seurat()</a></code> 
        </dt>
        <dd>Create Seurat Object with Cell Bender and Raw data</dd>
      </dl><dl><dt>
          
          <code><a href="Pull_Directory_List.html">Pull_Directory_List()</a></code> 
        </dt>
        <dd>Pull Directory List</dd>
      </dl><dl><dt>
          
          <code><a href="Read10X_GEO.html">Read10X_GEO()</a></code> 
        </dt>
        <dd>Load in NCBI GEO data from 10X</dd>
      </dl><dl><dt>
          
          <code><a href="Read10X_Multi_Directory.html">Read10X_Multi_Directory()</a></code> 
        </dt>
        <dd>Load 10X count matrices from multiple directories</dd>
      </dl><dl><dt>
          
          <code><a href="Read10X_h5_GEO.html">Read10X_h5_GEO()</a></code> 
        </dt>
        <dd>Load in NCBI GEO data from 10X in HDF5 file format</dd>
      </dl><dl><dt>
          
          <code><a href="Read10X_h5_Multi_Directory.html">Read10X_h5_Multi_Directory()</a></code> 
        </dt>
        <dd>Load 10X h5 count matrices from multiple directories</dd>
      </dl><dl><dt>
          
          <code><a href="Read_GEO_Delim.html">Read_GEO_Delim()</a></code> 
        </dt>
        <dd>Load in NCBI GEO data formatted as single file per sample</dd>
      </dl><dl><dt>
          
          <code><a href="Read_Metrics_10X.html">Read_Metrics_10X()</a></code> 
        </dt>
        <dd>Read Overall Statistics from 10X Cell Ranger Count</dd>
      </dl></div><div class="section level2">
      <h2 id="sequencing-metrics-qc-plots">Sequencing Metrics QC Plots<a class="anchor" aria-label="anchor" href="#sequencing-metrics-qc-plots"></a></h2>
      
      <p class="section-desc"></p><p>Functions plotting various QC metrics provides by 10X Genomics Cell Ranger Outputs</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="Seq_QC_Plot_Alignment_Combined.html">Seq_QC_Plot_Alignment_Combined()</a></code> 
        </dt>
        <dd>QC Plots Sequencing metrics (Alignment) (Layout)</dd>
      </dl><dl><dt>
          
          <code><a href="Seq_QC_Plot_Basic_Combined.html">Seq_QC_Plot_Basic_Combined()</a></code> 
        </dt>
        <dd>QC Plots Sequencing metrics (Layout)</dd>
      </dl><dl><dt>
          
          <code><a href="Seq_QC_Plot_Genes.html">Seq_QC_Plot_Genes()</a></code> 
        </dt>
        <dd>QC Plots Sequencing metrics</dd>
      </dl><dl><dt>
          
          <code><a href="Seq_QC_Plot_Number_Cells.html">Seq_QC_Plot_Number_Cells()</a></code> 
        </dt>
        <dd>QC Plots Sequencing metrics</dd>
      </dl><dl><dt>
          
          <code><a href="Seq_QC_Plot_Reads_in_Cells.html">Seq_QC_Plot_Reads_in_Cells()</a></code> 
        </dt>
        <dd>QC Plots Sequencing metrics</dd>
      </dl><dl><dt>
          
          <code><a href="Seq_QC_Plot_Reads_per_Cell.html">Seq_QC_Plot_Reads_per_Cell()</a></code> 
        </dt>
        <dd>QC Plots Sequencing metrics</dd>
      </dl><dl><dt>
          
          <code><a href="Seq_QC_Plot_Saturation.html">Seq_QC_Plot_Saturation()</a></code> 
        </dt>
        <dd>QC Plots Sequencing metrics</dd>
      </dl><dl><dt>
          
          <code><a href="Seq_QC_Plot_Total_Genes.html">Seq_QC_Plot_Total_Genes()</a></code> 
        </dt>
        <dd>QC Plots Sequencing metrics</dd>
      </dl><dl><dt>
          
          <code><a href="Seq_QC_Plot_UMIs.html">Seq_QC_Plot_UMIs()</a></code> 
        </dt>
        <dd>QC Plots Sequencing metrics</dd>
      </dl></div><div class="section level2">
      <h2 id="basic-analysis-qc-plots">Basic Analysis QC Plots<a class="anchor" aria-label="anchor" href="#basic-analysis-qc-plots"></a></h2>
      
      <p class="section-desc"></p><p>Functions plotting various QC from Seurat Objects.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="QC_Plot_GenevsFeature.html">QC_Plot_GenevsFeature()</a></code> 
        </dt>
        <dd>QC Plots Genes vs Misc</dd>
      </dl><dl><dt>
          
          <code><a href="QC_Plot_UMIvsFeature.html">QC_Plot_UMIvsFeature()</a></code> 
        </dt>
        <dd>QC Plots UMI vs Misc</dd>
      </dl><dl><dt>
          
          <code><a href="QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene()</a></code> 
        </dt>
        <dd>QC Plots Genes vs UMIs</dd>
      </dl><dl><dt>
          
          <code><a href="QC_Plots_Feature.html">QC_Plots_Feature()</a></code> 
        </dt>
        <dd>QC Plots Feature</dd>
      </dl><dl><dt>
          
          <code><a href="QC_Plots_Genes.html">QC_Plots_Genes()</a></code> 
        </dt>
        <dd>QC Plots Genes</dd>
      </dl><dl><dt>
          
          <code><a href="QC_Plots_Mito.html">QC_Plots_Mito()</a></code> 
        </dt>
        <dd>QC Plots Mito</dd>
      </dl><dl><dt>
          
          <code><a href="QC_Plots_UMIs.html">QC_Plots_UMIs()</a></code> 
        </dt>
        <dd>QC Plots UMIs</dd>
      </dl></div><div class="section level2">
      <h2 id="seurat-plotting-functions">Seurat Plotting Functions<a class="anchor" aria-label="anchor" href="#seurat-plotting-functions"></a></h2>
      
      <p class="section-desc"></p><p>Functions customization and plotting of single cell data/results from Seurat Objects.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="Cluster_Highlight_Plot.html">Cluster_Highlight_Plot()</a></code> 
        </dt>
        <dd>Cluster Highlight Plot</dd>
      </dl><dl><dt>
          
          <code><a href="DimPlot_All_Samples.html">DimPlot_All_Samples()</a></code> 
        </dt>
        <dd>DimPlot by Meta Data Column</dd>
      </dl><dl><dt>
          
          <code><a href="DimPlot_scCustom.html">DimPlot_scCustom()</a></code> 
        </dt>
        <dd>DimPlot with modified default settings</dd>
      </dl><dl><dt>
          
          <code><a href="DotPlot_scCustom.html">DotPlot_scCustom()</a></code> 
        </dt>
        <dd>Customized DotPlot</dd>
      </dl><dl><dt>
          
          <code><a href="FeaturePlot_DualAssay.html">FeaturePlot_DualAssay()</a></code> 
        </dt>
        <dd>Customize FeaturePlot of two assays</dd>
      </dl><dl><dt>
          
          <code><a href="FeaturePlot_scCustom.html">FeaturePlot_scCustom()</a></code> 
        </dt>
        <dd>Customize FeaturePlot</dd>
      </dl><dl><dt>
          
          <code><a href="Meta_Highlight_Plot.html">Meta_Highlight_Plot()</a></code> 
        </dt>
        <dd>Meta Highlight Plot</dd>
      </dl><dl><dt>
          
          <code><a href="PC_Plotting.html">PC_Plotting()</a></code> 
        </dt>
        <dd>PC Plots</dd>
      </dl><dl><dt>
          
          <code><a href="Split_FeatureScatter.html">Split_FeatureScatter()</a></code> 
        </dt>
        <dd>Split FeatureScatter</dd>
      </dl><dl><dt>
          
          <code><a href="Stacked_VlnPlot.html">Stacked_VlnPlot()</a></code> 
        </dt>
        <dd>Stacked Violin Plot</dd>
      </dl><dl><dt>
          
          <code><a href="VariableFeaturePlot_scCustom.html">VariableFeaturePlot_scCustom()</a></code> 
        </dt>
        <dd>Custom Labeled Variable Features Plot</dd>
      </dl><dl><dt>
          
          <code><a href="Plot_Density_Custom.html">Plot_Density_Custom()</a></code> 
        </dt>
        <dd>Nebulosa Density Plot</dd>
      </dl><dl><dt>
          
          <code><a href="Plot_Density_Joint_Only.html">Plot_Density_Joint_Only()</a></code> 
        </dt>
        <dd>Nebulosa Joint Density Plot</dd>
      </dl></div><div class="section level2">
      <h2 id="iterative-plotting-functions">Iterative Plotting Functions<a class="anchor" aria-label="anchor" href="#iterative-plotting-functions"></a></h2>
      
      <p class="section-desc"></p><p>Functions for quick iterative plotting.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="Iterate_Cluster_Highlight_Plot.html">Iterate_Cluster_Highlight_Plot()</a></code> 
        </dt>
        <dd>Iterate Cluster Highlight Plot</dd>
      </dl><dl><dt>
          
          <code><a href="Iterate_DimPlot_bySample.html">Iterate_DimPlot_bySample()</a></code> 
        </dt>
        <dd>Iterate DimPlot By Sample</dd>
      </dl><dl><dt>
          
          <code><a href="Iterate_FeaturePlot_scCustom.html">Iterate_FeaturePlot_scCustom()</a></code> 
        </dt>
        <dd>Iterative Plotting of Gene Lists using Custom Featureplots</dd>
      </dl><dl><dt>
          
          <code><a href="Iterate_Meta_Highlight_Plot.html">Iterate_Meta_Highlight_Plot()</a></code> 
        </dt>
        <dd>Iterate Meta Highlight Plot</dd>
      </dl><dl><dt>
          
          <code><a href="Iterate_PC_Loading_Plots.html">Iterate_PC_Loading_Plots()</a></code> 
        </dt>
        <dd>Iterate PC Loading Plots</dd>
      </dl><dl><dt>
          
          <code><a href="Iterate_Plot_Density_Custom.html">Iterate_Plot_Density_Custom()</a></code> 
        </dt>
        <dd>Iterative Plotting of Gene Lists using Custom Density Plots</dd>
      </dl><dl><dt>
          
          <code><a href="Iterate_Plot_Density_Joint.html">Iterate_Plot_Density_Joint()</a></code> 
        </dt>
        <dd>Iterative Plotting of Gene Lists using Custom Joint Density Plots</dd>
      </dl><dl><dt>
          
          <code><a href="Iterate_VlnPlot.html">Iterate_VlnPlot()</a></code> 
        </dt>
        <dd>Iterative Plotting of Gene Lists using VlnPlots</dd>
      </dl></div><div class="section level2">
      <h2 id="liger-plotting-functions">LIGER Plotting Functions<a class="anchor" aria-label="anchor" href="#liger-plotting-functions"></a></h2>
      
      <p class="section-desc"></p><p>Functions customization and plotting of single cell data/results from LIGER Objects.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="DimPlot_LIGER.html">DimPlot_LIGER()</a></code> 
        </dt>
        <dd>DimPlot LIGER Version</dd>
      </dl><dl><dt>
          
          <code><a href="plotFactors_scCustom.html">plotFactors_scCustom()</a></code> 
        </dt>
        <dd>Customized version of plotFactors</dd>
      </dl></div><div class="section level2">
      <h2 id="plotting-utilities-color-palettes-">Plotting Utilities #1 (Color Palettes)<a class="anchor" aria-label="anchor" href="#plotting-utilities-color-palettes-"></a></h2>
      
      <p class="section-desc"></p><p>Helper functions for quick access to useful or default color palettes.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="ColorBlind_Pal.html">ColorBlind_Pal()</a></code> 
        </dt>
        <dd>Color Universal Design Short Palette</dd>
      </dl><dl><dt>
          
          <code><a href="Dark2_Pal.html">Dark2_Pal()</a></code> 
        </dt>
        <dd>Dark2 Palette</dd>
      </dl><dl><dt>
          
          <code><a href="DiscretePalette_scCustomize.html">DiscretePalette_scCustomize()</a></code> 
        </dt>
        <dd>Discrete colour palettes from the pals package</dd>
      </dl><dl><dt>
          
          <code><a href="Hue_Pal.html">Hue_Pal()</a></code> 
        </dt>
        <dd>Hue_Pal</dd>
      </dl><dl><dt>
          
          <code><a href="JCO_Four.html">JCO_Four()</a></code> 
        </dt>
        <dd>Four Color Palette (JCO)</dd>
      </dl><dl><dt>
          
          <code><a href="NavyAndOrange.html">NavyAndOrange()</a></code> 
        </dt>
        <dd>Navy and Orange Dual Color Palette</dd>
      </dl><dl><dt>
          
          <code><a href="PalettePlot.html">PalettePlot()</a></code> 
        </dt>
        <dd>Plot color palette in viewer</dd>
      </dl><dl><dt>
          
          <code><a href="Single_Color_Palette.html">Single_Color_Palette()</a></code> 
        </dt>
        <dd>Single Color Palettes for Plotting</dd>
      </dl><dl><dt>
          
          <code><a href="scCustomize_Palette.html">scCustomize_Palette()</a></code> 
        </dt>
        <dd>Color Palette Selection for scCustomize</dd>
      </dl><dl><dt>
          
          <code><a href="viridis_shortcut.html">viridis_plasma_dark_high</a></code> <code><a href="viridis_shortcut.html">viridis_plasma_light_high</a></code> <code><a href="viridis_shortcut.html">viridis_inferno_dark_high</a></code> <code><a href="viridis_shortcut.html">viridis_inferno_light_high</a></code> <code><a href="viridis_shortcut.html">viridis_magma_dark_high</a></code> <code><a href="viridis_shortcut.html">viridis_magma_light_high</a></code> <code><a href="viridis_shortcut.html">viridis_dark_high</a></code> <code><a href="viridis_shortcut.html">viridis_light_high</a></code> 
        </dt>
        <dd>Viridis Shortcuts</dd>
      </dl></div><div class="section level2">
      <h2 id="plotting-utilities-themes-theme-adjustment-">Plotting Utilities #2 (Themes/Theme Adjustment)<a class="anchor" aria-label="anchor" href="#plotting-utilities-themes-theme-adjustment-"></a></h2>
      
      <p class="section-desc"></p><p>Helper functions for quick modification of theme elements and custom themes.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="Blank_Theme.html">Blank_Theme()</a></code> 
        </dt>
        <dd>Blank Theme</dd>
      </dl><dl><dt>
          
          <code><a href="Move_Legend.html">Move_Legend()</a></code> 
        </dt>
        <dd>Move Legend Position</dd>
      </dl><dl><dt>
          
          <code><a href="UnRotate_X.html">UnRotate_X()</a></code> 
        </dt>
        <dd>Unrotate x axis on VlnPLot</dd>
      </dl><dl><dt>
          
          <code><a href="theme_ggprism_mod.html">theme_ggprism_mod()</a></code> 
        </dt>
        <dd>Modified ggprism theme</dd>
      </dl></div><div class="section level2">
      <h2 id="helper-utilities-seurat-">Helper Utilities (Seurat)<a class="anchor" aria-label="anchor" href="#helper-utilities-seurat-"></a></h2>
      
      <p class="section-desc"></p><p>Functions to provide ease of use for frequently used code from Seurat Objects.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="Case_Check.html">Case_Check()</a></code> 
        </dt>
        <dd>Check for alternate case features
Checks Seurat object for the presence of features with the same spelling but alternate case.</dd>
      </dl><dl><dt>
          
          <code><a href="Change_Delim_All.html">Change_Delim_All()</a></code> 
        </dt>
        <dd>Change all delimiters in cell name</dd>
      </dl><dl><dt>
          
          <code><a href="Change_Delim_Prefix.html">Change_Delim_Prefix()</a></code> 
        </dt>
        <dd>Change barcode prefix delimiter</dd>
      </dl><dl><dt>
          
          <code><a href="Change_Delim_Suffix.html">Change_Delim_Suffix()</a></code> 
        </dt>
        <dd>Change barcode suffix delimiter</dd>
      </dl><dl><dt>
          
          <code><a href="Gene_Present.html">Gene_Present()</a></code> 
        </dt>
        <dd>Check if genes/features are present</dd>
      </dl><dl><dt>
          
          <code><a href="Merge_Sparse_Data_All.html">Merge_Sparse_Data_All()</a></code> 
        </dt>
        <dd>Merge a list of Sparse Matrices</dd>
      </dl><dl><dt>
          
          <code><a href="Meta_Numeric.html">Meta_Numeric()</a></code> 
        </dt>
        <dd>Check if meta data columns are numeric</dd>
      </dl><dl><dt>
          
          <code><a href="Meta_Present.html">Meta_Present()</a></code> 
        </dt>
        <dd>Check if meta data are present</dd>
      </dl><dl><dt>
          
          <code><a href="Replace_Suffix.html">Replace_Suffix()</a></code> 
        </dt>
        <dd>Replace barcode suffixes</dd>
      </dl><dl><dt>
          
          <code><a href="Add_Cell_Bender_Diff.html">Add_Cell_Bender_Diff()</a></code> 
        </dt>
        <dd>Calculate and add differences post-cell bender analysis</dd>
      </dl><dl><dt>
          
          <code><a href="Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat()</a></code> 
        </dt>
        <dd>Add Mito and Ribo percentages</dd>
      </dl><dl><dt>
          
          <code><a href="Liger_to_Seurat.html">Liger_to_Seurat()</a></code> 
        </dt>
        <dd>Create a Seurat object containing the data from a liger object</dd>
      </dl><dl><dt>
          
          <code><a href="Merge_Seurat_List.html">Merge_Seurat_List()</a></code> 
        </dt>
        <dd>Merge a list of Seurat Objects</dd>
      </dl><dl><dt>
          
          <code><a href="Meta_Remove_Seurat.html">Meta_Remove_Seurat()</a></code> 
        </dt>
        <dd>Remove meta data columns containing Seurat Defaults</dd>
      </dl><dl><dt>
          
          <code><a href="Rename_Clusters.html">Rename_Clusters()</a></code> 
        </dt>
        <dd>Rename Cluster Seurat</dd>
      </dl><dl><dt>
          
          <code><a href="Store_Misc_Info_Seurat.html">Store_Misc_Info_Seurat()</a></code> 
        </dt>
        <dd>Store misc data in Seurat object</dd>
      </dl><dl><dt>
          
          <code><a href="Store_Palette_Seurat.html">Store_Palette_Seurat()</a></code> 
        </dt>
        <dd>Store color palette in Seurat object</dd>
      </dl></div><div class="section level2">
      <h2 id="helper-utilities-liger-">Helper Utilities (LIGER)<a class="anchor" aria-label="anchor" href="#helper-utilities-liger-"></a></h2>
      
      <p class="section-desc"></p><p>Functions to provide ease of use for frequently used code from LIGER Objects.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="Add_Mito_Ribo_LIGER.html">Add_Mito_Ribo_LIGER()</a></code> 
        </dt>
        <dd>Add Mito and Ribo percentages to LIGER</dd>
      </dl><dl><dt>
          
          <code><a href="Meta_Present_LIGER.html">Meta_Present_LIGER()</a></code> 
        </dt>
        <dd>Check if meta data are present</dd>
      </dl><dl><dt>
          
          <code><a href="Top_Genes_Factor.html">Top_Genes_Factor()</a></code> 
        </dt>
        <dd>Extract top loading genes for LIGER factor</dd>
      </dl></div><div class="section level2">
      <h2 id="cluster-marker-annotation-utilities">Cluster/Marker Annotation Utilities<a class="anchor" aria-label="anchor" href="#cluster-marker-annotation-utilities"></a></h2>
      
      <p class="section-desc"></p><p>Functions to provide ease of use or add functionality to DEG and cluster annotation.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="Add_Pct_Diff.html">Add_Pct_Diff()</a></code> 
        </dt>
        <dd>Add percentage difference to DE results</dd>
      </dl><dl><dt>
          
          <code><a href="Create_Cluster_Annotation_File.html">Create_Cluster_Annotation_File()</a></code> 
        </dt>
        <dd>Create cluster annotation csv file</dd>
      </dl><dl><dt>
          
          <code><a href="Extract_Top_Markers.html">Extract_Top_Markers()</a></code> 
        </dt>
        <dd>Extract Top N Marker Genes</dd>
      </dl><dl><dt>
          
          <code><a href="Pull_Cluster_Annotation.html">Pull_Cluster_Annotation()</a></code> 
        </dt>
        <dd>Pull cluster information from annotation csv file.</dd>
      </dl></div><div class="section level2">
      <h2 id="project-organization-utilities">Project Organization Utilities<a class="anchor" aria-label="anchor" href="#project-organization-utilities"></a></h2>
      
      <p class="section-desc"></p><p>Functions to provide ease of use for organization of analysis projects.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="Copy_From_GCP.html">Copy_From_GCP()</a></code> 
        </dt>
        <dd>Copy folder from GCP bucket from R Console</dd>
      </dl><dl><dt>
          
          <code><a href="Copy_To_GCP.html">Copy_To_GCP()</a></code> 
        </dt>
        <dd>Copy folder to GCP bucket from R Console</dd>
      </dl><dl><dt>
          
          <code><a href="Setup_scRNAseq_Project.html">Setup_scRNAseq_Project()</a></code> 
        </dt>
        <dd>Setup project directory structure</dd>
      </dl></div><div class="section level2">
      <h2 id="statistics-functions">Statistics Functions<a class="anchor" aria-label="anchor" href="#statistics-functions"></a></h2>
      
      <p class="section-desc"></p><p>Functions quick return of different object and data metrics.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="Cluster_Stats_All_Samples.html">Cluster_Stats_All_Samples()</a></code> 
        </dt>
        <dd>Calculate Cluster Stats</dd>
      </dl><dl><dt>
          
          <code><a href="Median_Stats.html">Median_Stats()</a></code> 
        </dt>
        <dd>Median Statistics</dd>
      </dl><dl><dt>
          
          <code><a href="Percent_Expressing.html">Percent_Expressing()</a></code> 
        </dt>
        <dd>Calculate percent of expressing cells</dd>
      </dl><dl><dt>
          
          <code><a href="Plot_Cells_per_Sample.html">Plot_Cells_per_Sample()</a></code> 
        </dt>
        <dd>Plot Number of Cells/Nuclei per Sample</dd>
      </dl><dl><dt>
          
          <code><a href="Plot_Median_Genes.html">Plot_Median_Genes()</a></code> 
        </dt>
        <dd>Plot Median Genes per Cell per Sample</dd>
      </dl><dl><dt>
          
          <code><a href="Plot_Median_Mito.html">Plot_Median_Mito()</a></code> 
        </dt>
        <dd>Plot Median Percent Mito per Cell per Sample</dd>
      </dl><dl><dt>
          
          <code><a href="Plot_Median_Other.html">Plot_Median_Other()</a></code> 
        </dt>
        <dd>Plot Median other variable per Cell per Sample</dd>
      </dl><dl><dt>
          
          <code><a href="Plot_Median_UMIs.html">Plot_Median_UMIs()</a></code> 
        </dt>
        <dd>Plot Median UMIs per Cell per Sample</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.0.</p>
</div>

    </footer></div>

  

  

  </body></html>

